subprojects/core/src/main/java/org/gradle/api/internal/DefaultDomainObjectCollection.java

```diff
@@ -289,7 +291,14 @@ public void addLater(Provider<? extends T> provider) {
     public void addAllLater(Provider<? extends Iterable<T>> provider) {
         assertMutable("addAllLater(Provider)");
         assertMutableCollectionContents();
-        CollectionProviderInternal<T, ? extends Iterable<T>> providerInternal = Cast.uncheckedCast(provider);
+        final CollectionProviderInternal<T, ? extends Iterable<T>> providerInternal;
+        if (provider instanceof CollectionProviderInternal) {
+            providerInternal = Cast.uncheckedCast(provider);
+        } else {
+            // We don't know the type of element in the provider, so we assume it's the type of the collection
+            DefaultListProperty<T> defaultListProperty = new DefaultListProperty<T>(PropertyHost.NO_OP, Cast.uncheckedCast(getType()));
+            providerInternal = Cast.uncheckedCast(defaultListProperty.convention(provider));
```

Comment:
Why is a cast necessary here?  `DefaultListProperty` is-a `CollectionPropertyInternal`.

Reply:
I split this up, it was because `convention` returns a `ListProperty` which isn't a `CollectionProviderInternal`.